src/dmd/expression.d

```diff
@@ -7073,9 +7073,17 @@ extern (C++) final class FileInitExp : DefaultInitExp
     override Expression resolveLoc(Loc loc, Scope* sc)
     {
         //printf("FileInitExp::resolve() %s\n", toChars());
-        const(char)* s = loc.isValid() ? loc.filename : sc._module.ident.toChars();
+        const(char)* s;
         if (subop == TOK.fileFullPath)
-            s = FileName.combine(sc._module.srcfilePath, s);
+        {
+            if (loc.isValid())
+                s = FileName.toAbsolute(loc.filename);
```

Comment:
I'm not sure that `getcwd` will work in the general case. With `-i` and `-I` the file could be anywhere. As a start, we might want to add a test for more complicated cases?

Reply:
Ok, I've refactored the change a bit.  I've added a couple new fields to `Loc`.  One called `path` that can hold a path to `filename`.  Note that the path can be `null` if `filename` is absolute or relative to `CWD`.

I also added a field called `filenameType` which tracks whether `filename` came from an actual file or was generated from a `mixin` or `#line`.  This makes the code simpler in general for `__FILE_FULL_PATH__` because now it only generates an absolute path if the name represents a real file.

Reply:
Ok, we've gone back to the original implementation.  @ibuclaw has assured me that `filename` will always either be absolute or relative to CWD.  He says this MUST be the case otherwise it will break the debugger.  Also, since I couldn't find a counter-example it sounds like a reasonable thing to assume for now.  If we ever do find a counter-example we can address it later (with a test!).